src/main/java/com/commercetools/sync/producttypes/utils/ProductTypeReferenceReplacementUtils.java

```diff
@@ -0,0 +1,240 @@
+package com.commercetools.sync.producttypes.utils;
+
+import com.commercetools.sync.commons.exceptions.InvalidAttributeDefinitionException;
+import com.commercetools.sync.commons.exceptions.InvalidProductTypeException;
+import com.commercetools.sync.commons.exceptions.InvalidReferenceException;
+import com.commercetools.sync.commons.exceptions.ReferenceReplacementException;
+import io.sphere.sdk.expansion.ExpansionPath;
+import io.sphere.sdk.models.Reference;
+import io.sphere.sdk.products.attributes.AttributeDefinitionDraft;
+import io.sphere.sdk.products.attributes.AttributeDefinitionDraftBuilder;
+import io.sphere.sdk.products.attributes.AttributeType;
+import io.sphere.sdk.products.attributes.NestedAttributeType;
+import io.sphere.sdk.products.attributes.SetAttributeType;
+import io.sphere.sdk.producttypes.ProductType;
+import io.sphere.sdk.producttypes.ProductTypeDraft;
+import io.sphere.sdk.producttypes.ProductTypeDraftBuilder;
+import io.sphere.sdk.producttypes.queries.ProductTypeQuery;
+
+import javax.annotation.Nonnull;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+
+public final class ProductTypeReferenceReplacementUtils {
+
+    /**
+     * Takes a list of ProductTypes that are supposed to have their productType references (in case it has nestedType or
+     * set of NestedType) references expanded in order to be able to fetch the keys and replace the reference ids with
+     * the corresponding keys and then return a new list of productType drafts with their references containing keys
+     * instead of the ids.
+     *
+     * <p><b>Note:</b>If some references are not expanded for an attributeDefinition of a productType, the method will
+     * throw a {@link ReferenceReplacementException} containing the root causes of the exceptions that occurred in any
+     * of the supplied {@code productTypes}.
+     *
+     * @param productTypes the list of productTypes to to replace the references on and convert to productTypeDrafts.
+     * @return a list of productType drafts with keys instead of ids for references.
+     */
+    @Nonnull
+    public static List<ProductTypeDraft> replaceProductTypesReferenceIdsWithKeys(
+        @Nonnull final List<ProductType> productTypes) {
+
+        final Set<Throwable> potentialErrors = new HashSet<>();
```

Comment:
what do you mean by `potantialErrors` ?  maybe use `referenceReplacementErrors` ?

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/402#discussion_r301514507